### PR TITLE
Added table oci_containerengine_cluster. Closes #383

### LINKED ISSUE #383: Add table oci_containerengine_cluster
Required for reporting

**References**
https://registry.terraform.io/providers/oracle/oci/latest/docs/resources/containerengine_cluster

### PR DESCRIPTION
**Pull request by karanpopat**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
Outputs:

lifecycle_state = "ACTIVE"
resource_id = "ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaa2hxj3vjhpsgcuvqkqwtfci3eaq3z5jq3ls3c2t4ovc4t45n7fiwq"
resource_name = "steampipetest3066"
tenancy_ocid = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"

Running SQL query: test-get-query.sql

[
  {
    "id": "ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaa2hxj3vjhpsgcuvqkqwtfci3eaq3z5jq3ls3c2t4ovc4t45n7fiwq",
    "lifecycle_state": "ACTIVE",
    "name": "steampipetest3066"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql

[
  {
    "id": "ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaa2hxj3vjhpsgcuvqkqwtfci3eaq3z5jq3ls3c2t4ovc4t45n7fiwq",
    "lifecycle_state": "ACTIVE",
    "name": "steampipetest3066"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql

null
✔ PASSED

Running SQL query: test-turbot-query.sql

[
  {
    "tenant_id": "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq",
    "title": "steampipetest3066"
  }
]
✔ PASSED

POSTTEST: tests/oci_containerengine_cluster

TEARDOWN: tests/oci_containerengine_cluster

SUMMARY:

1/1 passed.

```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  name,
  id,
  lifecycle_state
from
  oci_containerengine_cluster;
+-------------------+--------------------------------------------------------------------------------------------+-----------------+
| name              | id                                                                                         | lifecycle_state |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+
| cluster1          | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaagrhzevmfhatblxq6qeeql4i4oljqcijfuk6g7iucbc5tm4wzor4a | DELETED         |
| steampipetest1030 | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaahpwphy5bipnzq7fmkuaynftleqp637inh5xirkt2qc7nyke5umva | CREATING        |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+

> select
  name,
  id,
  lifecycle_state
from
  oci_containerengine_cluster
where
  lifecycle_state <> 'AVAILABLE';
+-------------------+--------------------------------------------------------------------------------------------+-----------------+
| name              | id                                                                                         | lifecycle_state |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+
| cluster1          | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaagrhzevmfhatblxq6qeeql4i4oljqcijfuk6g7iucbc5tm4wzor4a | DELETED         |
| steampipetest1030 | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaahpwphy5bipnzq7fmkuaynftleqp637inh5xirkt2qc7nyke5umva | CREATING        |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+

> select
  name,
  id,
  lifecycle_state,
  image_policy_config_enabled
from
  oci_containerengine_cluster
where
  image_policy_config_enabled = false;
+-------------------+--------------------------------------------------------------------------------------------+-----------------+-----------------------------+
| name              | id                                                                                         | lifecycle_state | image_policy_config_enabled |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+-----------------------------+
| steampipetest1030 | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaahpwphy5bipnzq7fmkuaynftleqp637inh5xirkt2qc7nyke5umva | CREATING        | false                       |
| cluster1          | ocid1.cluster.oc1.ap-mumbai-1.aaaaaaaagrhzevmfhatblxq6qeeql4i4oljqcijfuk6g7iucbc5tm4wzor4a | DELETED         | false                       |
+-------------------+--------------------------------------------------------------------------------------------+-----------------+-----------------------------+
```
</details>

### CODE REVIEW

File: docs/tables/oci_containerengine_cluster.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+```sql
+select
+  name,
+  id,
+  lifecycle_state
+from
+  oci_containerengine_cluster;
+```
+
+### List Container Engine Clusters that are not available
```
Comment (ParthaI) [inline docs/tables/oci_containerengine_cluster.md:18]:
```suggestion
### List unavailable container engine clusters 
```

File: oci-test/tests/oci_containerengine_cluster/variables.tf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+variable "cluster_kubernetes_version" {
+  type        = string
+  default     = "v1.22.5"
+  description = "The version of Kubernetes to install into the cluster masters."
+}
+
+variable "tenancy_ocid" {
+  type        = string
+  default     = "ocid1.tenancy.oc1..aaaaaaaahnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4h2q37cyljaq"
```
Comment (ParthaI) [inline oci-test/tests/oci_containerengine_cluster/variables.tf:21]:
```suggestion
  default     = ""
```

File: oci/table_oci_containerengine_cluster.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+			},
+			{
+				Name:        "id",
+				Description: "The OCID of the cluster.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromCamel(),
+			},
+			{
+				Name:        "lifecycle_state",
+				Description: "The state of the cluster masters",
```
Comment (ParthaI) [inline oci/table_oci_containerengine_cluster.go:59]:
```suggestion
				Description: "The state of the cluster masters.",
```

**Review by ParthaI — COMMENTED** [review]
@karanpopat left few comments please have a look, can you add the optional quals example query result in pr?

File: oci/table_oci_containerengine_cluster.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+		return nil, err
+	}
+	return response.Cluster, nil
+}
+
+// Build additional filters
+func buildContainerEngineClusterFilters(equalQuals plugin.KeyColumnEqualsQualMap, logger hclog.Logger) (containerengine.ListClustersRequest, bool) {
+	request := containerengine.ListClustersRequest{}
+	isValid := true
+	if equalQuals["name"] != nil {
```
Comment (ParthaI) [inline oci/table_oci_containerengine_cluster.go:263]: Add extra check for empty value check, If you will try the below query then you will get the InvalidParameterException.

`elect * from oci_containerengine_cluster where name = ' '` or `elect * from oci_containerengine_cluster where name = '<more than 1 empty space>' ` or `elect * from oci_containerengine_cluster where name = ''`

File: docs/tables/oci_containerengine_cluster.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# Table: oci_containerengine_cluster
+
+Oracle Cloud Infrastructure Container Engine for Kubernetes is a fully-managed, scalable, and highly available service that you can use to deploy your containerized applications to the cloud
```
Comment (bigdatasourav) [inline docs/tables/oci_containerengine_cluster.md:3]:
```suggestion
Oracle Cloud Infrastructure Container Engine for Kubernetes is a fully-managed, scalable, and highly available service that you can use to deploy your containerized applications to the cloud.
```

File: docs/tables/oci_containerengine_cluster.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+```sql
+select
+  name,
+  id,
+  lifecycle_state
+from
+  oci_containerengine_cluster;
+```
+
+### List unavailable container engine clusters
```
Comment (bigdatasourav) [inline docs/tables/oci_containerengine_cluster.md:18]: List failed ones

File: oci/plugin.go [hunk 7]
```diff
@@ -38,6 +38,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"oci_cloud_guard_managed_list":                                 tableCloudGuardManagedList(ctx),
 			"oci_cloud_guard_responder_recipe":                             tableCloudGuardResponderRecipe(ctx),
 			"oci_cloud_guard_target":                                       tableCloudGuardTarget(ctx),
+			"oci_containerengine_cluster":                                  tableContainerEngineCluster(ctx),
```
Comment (bigdatasourav) [inline oci/plugin.go:41]:
```suggestion
			"oci_containerengine_cluster":                                  tableOciContainerEngineCluster(ctx),
```

File: oci/table_oci_containerengine_cluster.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+//// HYDRATE FUNCTION
+
+func getContainerEngineCluster(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	logger := plugin.Logger(ctx)
+	region := plugin.GetMatrixItem(ctx)[matrixKeyRegion].(string)
+	compartment := plugin.GetMatrixItem(ctx)[matrixKeyCompartment].(string)
+	logger.Debug("getContainerEngineClusters", "Compartment", compartment, "OCI_REGION", region)
+
+	// Restrict the api call to only root compartment/ per region
+	if !strings.HasPrefix(compartment, "ocid1.tenancy.oc1") {
```
Comment (bigdatasourav) [inline oci/table_oci_containerengine_cluster.go:223]: This check should be under else part in [here](https://github.com/turbot/steampipe-plugin-oci/pull/386/files#diff-16eda5d1f777f196a6a85f45244105f0b0332bff67e17c5994173434319de29aR230)

File: oci/table_oci_containerengine_cluster.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+					Name:    "compartment_id",
+					Require: plugin.Optional,
+				},
+				{
+					Name:    "lifecycle_state",
+					Require: plugin.Optional,
+				},
+			},
+		},
+		GetMatrixItem: BuildCompartementZonalList,
```
Comment (bigdatasourav) [inline oci/table_oci_containerengine_cluster.go:44]: We can use BuildCompartementList here, as we are not using zone.

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Pull request merged**